rust-runtime/inlineable/src/constrained.rs

```diff
@@ -0,0 +1,15 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+pub(crate) trait Constrained {
+    type Unconstrained;
+}
+
+#[derive(Debug, Clone)]
+#[allow(dead_code)]
+pub(crate) enum MaybeConstrained<T: Constrained> {
```

Comment:
> This commit makes the sever deserializers work with unconstrained types
during request parsing, and only after the entire request is parsed are
constraints enforced.

If

```rust
trait Constrained {
    type Unconstrained; 
    type Error;

    fn constrain(unconstrained: Self::Unconstrained) -> Result<Self, (Self::Unconstrained, Self::Error)>;
}
```

or something similar. The constrain logic gets code generated in this impl.

Then we replaced `MaybeConstrained` with

```rust
enum ConstraintAttempt<T: Constrained> {
    Success(T),
    Failure {
        unconstrained: T::Unconstrained,
        error: T::ConstraintError
    }
}
```

which fails to deserialize exactly when `T::Unconstrained` fails to deserialize:

- `T::Unconstrained` fails to deserialize then return an error,
- `T::Unconstrained` fails to constrain return an `Ok(ConstraintAttempt::Failure { ... })`,
- `T::Unconstrained` does contrain, return an `Ok(Success(T))`.


For example something like this could live alongside `ConstraintAttempt`:

```rust
impl<T> FromStr for ConstraintAttempt<T> where T: Constrained, FromStr {
    type Error = <T as FromStr>::Error;

    fn from_str(value: &str) -> Result<Self, Self::Error> {
        let unconstrained = <T as Constrained>::Unconstrained::from_str(value)?;
        let value = match T::constrain(unconstrained) {
            Ok(ok) => Self::Success(ok),
            Err((unconstrained, error)) => Self::Failure { unconstrained, error }
        };
        Ok(value)
    }
}
```

Then we can "check" constraints during request parsing, but be lazy about collecting errors?

> and only after the entire request is parsed are
constraints enforced.

This then becomes just checking whether `ConstraintAttempt::Success` or not.

Does this make things more compact?

Reply:
> Does this make things more compact?

I don't see how. It's just swapping `MaybeConstrained` for `ConstrainAttempt`.

> `T::Unconstrained` fails to parse then return an error

I don't see how this can happen. At this point the JSON deserializer has already deserialized a valid node into a Rust data type e.g. a `Vec<String>`, now it's just a matter of whether it satisfies the modeled constraints. Or are you suggesting that `FromStr for ConstraintAttempt<T>` do wire format parsing _and_ deserialization at the same time?

> Then we can "check" constraints during request parsing, but be lazy about collecting errors?

* Lazy error collection is currently how it works: error collection is done when the builder is built, because that's where constraints are checked.
* A benefit of doing constraint checking during request deserialization is that we can "short-circuit" on e.g. the `length` trait if we reach the maximum number of allowed items. But I can't see how `ConstrainAttempt` helps here, since it also takes in the already deserialized value.

Reply:
Yeah what you're describing is having `impl<T> FromStr for ConstraintAttempt<T>` do parsing _and_ deserializing at the same time. It's a viable approach (serde is similar in that the deserializer drives a visitor that does the parsing to map onto serde's data model), but it's not how are code is currently architected. Imagine you had a constrained map holding deeply nested data structures. `impl<T> FromStr for ConstrainAttempt<T>` is now not as straightforward as your implementation where it just delegates to `<T as Constrained>::Unconstrained::from_str`. It now has to parse the JSON tokens for the map elements and delegate to the nested data structures' deserializers, then compose the results. This is essentially what our `JsonParserGenerator` is doing. Its logic could be reused in `impl<T> FromStr for ConstraintAttempt<T>` of course, but it's a non-negligible deviation from our architecture, where we have deserializing and parsing as decoupled concerns.

I do think it's a valuable approach though, since it allows us to more easily short-circuit in the case of `@length`, whereas if we ever want to do that right now, we'd have to trickle down that intent through the `JsonParserGenerator`'s logic.

If we're going to fundamentally change the deserializer's architecture, I'd rather invest in trying out implementing serde's traits though.